### ISSUE TITLE
[R/CLI] Sparse dataset taking forever to load/save_binary

### ISSUE THREAD

**Issue by billytcl** [microsoft/LightGBM#5478]
## Description
I am using the current version of lightGBM from CRAN, and cloned the github repo on Sept 1 for the CLI. FWIW I've only been using lightGBM since last Friday so this is new to me!

I have a tab-delimited text file, which consists of 2998153 columns (features) and ~500 rows (observations). It's about ~95-98% sparse. Missing entries are "NA" and the rest are numbers from 0 to 100. The vast majority are either 0 or 100, with a small fraction having numbers in between. I also have a smaller validation set, which only has 40 rows, but is also taking forever to load/save_binary. The objective is binary classification. The machine is just stuck at 100% CPU with no log messages.

For smaller datasets (eg. with a more restricted set of features ~160k), the dataset loads quickly. This doesn't seem to be a memory issue so I don't know what is happening. So far it's using <50GB of system memory and my machine has 1TB.

Am I doing something wrong here? It seems like the loading got exponentially worse, even with the smaller validation file when I tried to use save_binary. But ~160k -> 3M doesn't seem like that far of a leap. My goal is to see how well lightGBM can pick out relevant features for good prediction, since I had pre-filtered them down to ~160k using another method. I'd like to see whether lightGBM can do it on its own. If this many features/sparsity is too much for lightGBM to handle, can you recommend other methods to efficiently pre-filter/select features to try out for lightGBM?

## Reproducible example

Here's my CLI invocation: (save_binary.conf)
```
task = save_binary
data = [...the file...]
max_bins = 10
min_data = 1
verbosity = 10
is_enable_sparse = true
use_missing = true
zero_as_missing = false
two_round = true
```

and the relevant R code:
```
library(lightgbm)
library(data.table)
library(dplyr)
library(ggplot2)

dataset <- fread([... the file ...], data.table = F, header = F)

set.seed(i_seed)
sample_training <- seq(1,nrow(dataset)) %>% sample(nrow(dataset),replace = T) %>% unique()
sample_test <- setdiff(seq(1,nrow(dataset)), sample_training)

training_dataset <- lgb.Dataset(as.matrix(dataset[sample_training,2:ncol(dataset)]), label = dataset[sample_training,]$V1)
test_dataset <- list(test = lgb.Dataset.create.valid(training_dataset, as.matrix(dataset[sample_test,2:ncol(dataset)]), label = dataset[sample_test,]$V1))

validation <- fread([... a smaller file ...], data.table=F, header=F)
validation_dataset <- as.matrix(validation[,2:ncol(validation)])

training_props <- prop.table(table(dataset[sample_training,]$V1))


params <- list(
  task = "train",
  objective = "binary",
  metric = c("binary_logloss","auc","average_precision"),
  is_unbalance = T,
  verbosity = 10,
  metric_freq = 5,
  min_data = 1,
  max_bin = 10,  
  num_trees = 200, 
  learning_rate = 0.2, 
  num_leaves = 200, 
  tree_learner = "feature", 
  num_threads = 40,
  seed = 0,
  feature_fraction = 1, 
  bagging_freq = 1, 
  bagging_fraction = .65, 
  is_enable_sparse = T,
  use_missing = T,
  zero_as_missing = F,
  use_two_round_loading = F,
  is_save_binary_file = T,
  extra_trees = T,
  force_col_wise = T
)

model <- lgb.train(
  params = params,
  data = training_dataset,
  valids = test_dataset
)
```

## Environment info

LightGBM version or commit hash: 3.3.2.99

```shell

```
